File: fortunecards.client/src/app/models/pattern.ts
import { PagedResult } from './deck';

export type { PagedResult };

export interface PatternCard {
  id?: number;
  text: string;
  order: number;
  x: number;
  y: number;
  rotation: number;
}

/** A pattern card being edited: carries a stable client id for tracking + table targeting. */
export interface EditablePatternCard {
  id: string;
  text: string;
  order: number;
  x: number;
  y: number;
  rotation: number;
}

export interface Pattern {
  id: number;
  name: string;
  description: string | null;
  createdAt: string;
  cardCount?: number;
  cards?: PatternCard[];
  emoji: string;
  colorIndex: number;
  isPublic: boolean;
  isOwner: boolean;
  isFavorite: boolean;
  cardSizePercent?: number;
  tableHeightPercent?: number;
}

export interface CreatePatternPayload {
  name: string;
  description: string | null;
  emoji: string;
  colorIndex: number;
  isPublic: boolean;
}

export interface UpdatePatternPayload {
  name?: string;
  description?: string | null;
  emoji?: string;
  colorIndex?: number;
  isPublic?: boolean;
  cardSizePercent?: number;
  tableHeightPercent?: number;
}

File: FortuneCards.Server/Models/Pattern.cs
namespace FortuneCards.Server.Models
{
    public class Pattern
    {
        public int Id { get; set; }
        public required string Name { get; set; }
        public string? Description { get; set; }
        public string Emoji { get; set; } = "🔮";
        public int ColorIndex { get; set; } = 0;
        public bool IsPublic { get; set; } = false;
        public int CardSizePercent { get; set; } = 15;
        public int TableHeightPercent { get; set; } = 60;
        public DateTime CreatedAt { get; set; } = DateTime.UtcNow;
        public int? UserId { get; set; }
        public User? User { get; set; }
        public ICollection<PatternCard> Cards { get; set; } = new List<PatternCard>();
        public ICollection<FavoritePattern> FavoritedBy { get; set; } = new List<FavoritePattern>();
    }
}
